### PR TITLE
macOS controller duplication bugfix

### PR DESCRIPTION
**Pull request by ComradeEcho**
## Description

Hey y'all. 

Just to preface, I have ~zero experience with github, and this is literally my first time ever touching C or anything close to this low-level. Please scrutinize the heck out of this, because I don't know what I'm doing. Especially pointer/memory/free()/deinit stuff.

This PR aims to take care of a couple of things having to do with the macOS HID Driver.

1) The `IOHIDDeviceRegisterRemovalCallback` simply was not working as I believe it should. I don't know if it was implimented improperly, or if there's a bug in macOS related to it. It was being set, but it wouldn't trigger reliably. I replaced `IOHIDDeviceRegisterRemovalCallback` with `IOHIDManagerRegisterDeviceMatchingCallback` which seemed to have perfect detection of device removal on my system.

2) Upon fixing the first issue, I discovered that devices matched by the autoconfigurator still were not disconnecting properly. This was caused an `struct iohidmanager_hid_adaptor` being improperly cast as a `struct pad_connection`, and having its data overwritten/corrupted.
I solved this issue by adding a `void* connection` to `joypad_connection` and assigning the offending `struct pad_connection` there instead of it going in `void* data`, thus removing the data collision.

3) I removed all references to `uniqueId` because I *assumed* it accomplished nothing. I later realized what it was accomplishing, but decided to try to implement the same goal with `kIOHIDLocationIDKey` instead, because `kIOHIDUniqueIDKey` is unsupported by powerPC
What it was accomplishing was: controllers are detected both when the application is launched, and by the controller connection callback. uniqueId was being used to make sure the controllers are not duplicated by getting added by both of those. 

4) ~~I made a two (unrelated) changes to how debug logs are shows in the logs. I don't know how to exclude those from my PR, but it fixes what I assume is a bug. (debug-level logs would *output* as info-level logs) :(~~

## Notes

I have done testing solely on M1 macOS 11.

I have done testing with these controllers: 
* xbox series s (bluetooth)
* ps3 (bluetooth)
* ps4 (wired and bluetooth)

I have access to a few more untested controllers including:
* Mayflash 
* 360
* joyconns
* switch pro
* switch-gamecube
* n64 usb adapter?
* maybe many more!

I have *absolutely* no idea if this affects windows/linux/android. It seems like a very Apple-heavy driver, and I don't know if the HID input driver is even functional on other platforms.

## Related Issues

For sure:
https://github.com/libretro/RetroArch/issues/12561
https://github.com/libretro/RetroArch/issues/11294
https://github.com/libretro/RetroArch/issues/10546

Possibly:
https://github.com/libretro/RetroArch/issues/3414
https://github.com/libretro/RetroArch/issues/2808
https://github.com/libretro/RetroArch/issues/4215
https://github.com/libretro/RetroArch/issues/3415
https://github.com/libretro/RetroArch/issues/6731
https://github.com/libretro/RetroArch/issues/4793

## Reviewers

@gblues @jdgleaver @hunterk @markwkidd 

@twinaphex With the changes made surrounding uniqueId - I think this needs a good test on powerPC

### CODE REVIEW

**Comment by ComradeEcho** [comment]
It was noted to me that there's quite a few whitespace inconsistencies. That wasn't something I could see while working on this in xcode, but will need to get fixed. There's also still a handful of debug `printf()` that will need to be removed as well. Although, perhaps some useful `RARCH_DBG()` could be added.

File: input/drivers_hid/iohidmanager_hid.c [hunk 1]
```diff
@@ -476,14 +474,43 @@ static void iohidmanager_hid_device_input_callback(void *data, IOReturn result,
...
+      else
+         printf("Port %d is assigned to device %p on HID %p \n", i, a->handle, hid);
+   }
+   printf("\n");
+}
+static void iohidmanager_hid_device_remove(IOHIDDeviceRef device, iohidmanager_hid_t* hid)
+{
+   struct iohidmanager_hid_adapter *adapter = NULL;
+   //loop though the controller ports and find the device with a matching IOHINDeviceRef
+   for (int i=0; i<MAX_USERS; i++)
```
Comment (twinaphex) [inline input/drivers_hid/iohidmanager_hid.c:499]: Can you change this? Don't use internal declaration for loops. Instead, define 'int i' outside of the for loop, at the start of a code function or block. This makes the code C89-compliant.

File: input/drivers_hid/iohidmanager_hid.c [hunk 2]
```diff
@@ -476,14 +474,43 @@ static void iohidmanager_hid_device_input_callback(void *data, IOReturn result,
...
+      }
+      else
+         printf("Port %d is assigned to device %p on HID %p \n", i, a->handle, hid);
+   }
+   printf("\n");
+}
+static void iohidmanager_hid_device_remove(IOHIDDeviceRef device, iohidmanager_hid_t* hid)
+{
+   struct iohidmanager_hid_adapter *adapter = NULL;
+   //loop though the controller ports and find the device with a matching IOHINDeviceRef
```
Comment (twinaphex) [inline input/drivers_hid/iohidmanager_hid.c:498]: Can you change these and other comments to C-based comments?

File: input/drivers_hid/iohidmanager_hid.c [hunk 3]
```diff
@@ -476,14 +474,43 @@ static void iohidmanager_hid_device_input_callback(void *data, IOReturn result,
...
+         continue;
+      if (a->handle == device)
+      {
+         adapter = a;
+         break;
+      }
+   }
+   if (!adapter)
+   {
+      printf("Error removing device %p from HID %p\n",device, hid);
```
Comment (twinaphex) [inline input/drivers_hid/iohidmanager_hid.c:512]: Are all these printfs necessary? We could replace them with RARCH_LOG/RARCH_ERR/RARCH_WARN if they are truly essential. Do note that the logging code is not thread-safe so if this function happens on a thread, perhaps not logging at all would be the safest choice. But I'll leave this determination up to you to decide.

**Review by twinaphex — CHANGES_REQUESTED** [review]

File: input/drivers_hid/iohidmanager_hid.c (hunk 3 above)
Comment (ComradeEcho) [reply input/drivers_hid/iohidmanager_hid.c:512]: The printfs aren't necessary, but I was reluctant to remove them yet as they may still be useful as things get tested.

However, I think *some* additional logs would be nice. As of right now the *only* log that's output is when a controller is connected. I think we at least have logs when a controller is disconnected as well as a log detailing whether the device was matched by the autoconfigurator.

I'm not sure where any of this falls in regards to threading. My untrained eye sees that this all seems to be going on within thread 1 - but I have no idea tbh. I'd love to learn about that :)
Screen Shot 2021-08-06 at 12 05 51 PM

**Comment by ComradeEcho** [comment]
I am currently getting a `SIGABRT` seemingly randomly while connecting/disconnecting controllers. I can't reproduce it. I can sit here connecting and reconnecting various controllers without incident, but every once in a while this will happen. It may be time-based.

`pointer being freed was not allocated` while trying to free a `iohidmanager_hid_adapter *adapter`

This is the point where I have no idea what the actual issue is, or how to solve it.

Screen Shot 2021-08-06 at 5 21 22 PM

File: input/drivers_hid/iohidmanager_hid.c (hunk 1 above)
Comment (ComradeEcho) [reply input/drivers_hid/iohidmanager_hid.c]: done

File: input/drivers_hid/iohidmanager_hid.c (hunk 2 above)
Comment (ComradeEcho) [reply input/drivers_hid/iohidmanager_hid.c]: done

File: input/drivers_hid/iohidmanager_hid.c (hunk 3 above)
Comment (ComradeEcho) [reply input/drivers_hid/iohidmanager_hid.c]: I removed the printfs, and added a couple logs, both info and debug

**Pull request closed without merging**